politeiad/backendv2/tstorebe/plugins/usermd/usermd.go

```diff
@@ -84,21 +86,145 @@ func (p *usermdPlugin) Hook(h plugins.HookT, payload string) error {
+	rs := make(map[string]*backend.Record, len(tokens))
+
+	// addMissingRecord adds the given record's token to a list of tokens sorted
+	// by the latest status change timestamp, from newest to oldest.
+	addMissingRecord := func(tokens []string, missingRecord *backend.Record) ([]string, error) {
+		// Make list of records to be able to sort by latest status change
+		// timestamp.
+		records := make([]*backend.Record, 0, len(tokens)+1)
+		for _, t := range tokens {
+			// Search in known records map
+			r := rs[t]
+
+			// Fetch record if not fetched yet
+			if r == nil {
+				// Decode string token
+				b, err := hex.DecodeString(t)
+				if err != nil {
+					return nil, err
+				}
+				r, err = p.tstore.RecordPartial(b, 0, nil, false)
+				if err != nil {
+					return nil, err
+				}
+
+				// Add record to the known records map
+				rs[t] = r
+			}
+
+			records = append(records, r)
+		}
+
+		// Append new record then sort reocrds by latest status change timestamp
+		// from newest to oldest.
+		records = append(records, missingRecord)
+
+		// Sort records
+		sort.Slice(records, func(i, j int) bool {
+			return records[i].RecordMetadata.Timestamp >
+				records[j].RecordMetadata.Timestamp
+		})
+
+		// Return sorted tokens
+		newTokens := make([]string, 0, len(records))
+		for _, record := range records {
+			newTokens = append(newTokens, record.RecordMetadata.Token)
+		}
+
+		return newTokens, nil
+	}
+
+	for _, token := range tokens {
+		// Check if record was already fetched
+		tokenStr := hex.EncodeToString(token)
+		r := rs[tokenStr]
+
+		// Fetch record if not fetched yet
+		if r == nil {
+			record, err := p.tstore.RecordPartial(token, 0, nil, false)
+			if err != nil {
+				return err
+			}
+
+			// Add record to the known records map
+			r = record
+			rs[tokenStr] = r
+		}
+
+		// Decode user metadata
+		um, err := userMetadataDecode(r.Metadata)
+		if err != nil {
+			return err
+		}
+
+		// Get the user cache for the record's author
+		uc, err := p.userCache(um.UserID)
+		if err != nil {
+			return err
+		}
+
+		// Verify that the record is listed in the user cache under the
+		// correct category.
+		var found bool
+		switch r.RecordMetadata.State {
+		case backend.StateUnvetted:
+			for _, t := range uc.Unvetted {
+				if t == hex.EncodeToString(token) {
+					found = true
+				}
+			}
+			// Unvetted record is missing, add it
+			if !found {
+				uc.Unvetted, err = addMissingRecord(uc.Unvetted, r)
+				if err != nil {
+					return err
+				}
+			}
+
+		case backend.StateVetted:
+			for _, t := range uc.Vetted {
+				if t == hex.EncodeToString(token) {
+					found = true
+				}
+			}
+			// Vetted record is missing, add it
+			if !found {
+				uc.Vetted, err = addMissingRecord(uc.Vetted, r)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// If a missing token was added to the user cache, save new user cache
+		// to disk.
+		if !found {
+			p.Lock()
+			err = p.userCacheSaveLocked(um.UserID, *uc)
```

Comment:
@lukebp should i log some info at this level, maybe:
`missing vetted/unvetted record $token was added to $userid user records cache`

Reply:
I would do high level actions/stats as `Infof` and individual records actions as `Debugf`.  Logging something for each record as `Infof` is too noisy.

Reply:
sweet, will do that, maybe will have a global counter and log how many missing records were added
and debugf here some info as i mentioned above

Reply:
done